### PR TITLE
Migrate SubmapTexture and SubmapSlice logics from cartographer_ros

### PR DESCRIPTION
**Pull request by jihoonl**
Towards [RFC06](https://github.com/googlecartographer/rfcs/blob/master/text/0006-serve-ros-map-from-pbstream.md).

Migrates
* `FillSubmapSlice` from `pbstream_to_rosmap_main.cc` 
* `SubmapTexture` logics from cartographer_ros

### CODE REVIEW

**Comment by cschuet** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user cschuet. Merge queue now has a length of 2.

File: cartographer/io/submap_painter.h [hunk 1]
```diff
@@ -55,10 +58,43 @@ struct SubmapSlice {
...
 
+void FillSubmapSlice(
+    const ::cartographer::transform::Rigid3d& global_submap_pose,
+    const ::cartographer::mapping::proto::Submap& proto,
+    SubmapSlice* const submap_slice);
+
+// Unpacks cell data as provided by the backend into 'intensity' and 'alpha'.
+SubmapTexture::Pixels UnpackTextureData(const std::string& compressed_cells,
+                                        int width, int height);
+
```
Comment (pifon2a) [inline cartographer/io/submap_painter.h]: In .cc file you have here 'const int width, const int height'. Same goes for DrawTexture(). Please, correct for consistency. Just remove "const"s, you are passing by value anyways.

File: cartographer/io/submap_painter.cc [hunk 2]
```diff
@@ -107,5 +104,82 @@ PaintSubmapSlicesResult PaintSubmapSlices(
   return PaintSubmapSlicesResult(std::move(surface), origin);
 }
 
+void FillSubmapSlice(
+    const ::cartographer::transform::Rigid3d& global_submap_pose,
+    const ::cartographer::mapping::proto::Submap& proto,
+    SubmapSlice* const submap_slice) {
+  ::cartographer::mapping::proto::SubmapQuery::Response response;
+  ::cartographer::transform::Rigid3d local_pose;
```
Comment (pifon2a) [inline cartographer/io/submap_painter.cc]: Should 'local_pose' be used after "submap.local_pose()" anywhere?

File: cartographer/io/submap_painter.cc [hunk 3]
```diff
@@ -107,5 +104,82 @@ PaintSubmapSlicesResult PaintSubmapSlices(
...
+    local_pose = submap.local_pose();
+    submap.ToResponseProto(global_submap_pose, &response);
+  } else {
+    ::cartographer::mapping_2d::Submap submap(proto.submap_2d());
+    local_pose = submap.local_pose();
+    submap.ToResponseProto(global_submap_pose, &response);
+  }
+  submap_slice->pose = global_submap_pose;
+
+  auto& texture_proto = response.textures(0);
```
Comment (pifon2a) [inline cartographer/io/submap_painter.cc]: should it be "const auto&"?

File: cartographer/io/submap_painter.h [hunk 4]
```diff
@@ -55,10 +58,43 @@ struct SubmapSlice {
   int metadata_version = -1;
 };
 
+struct SubmapTexture {
+  struct Pixels {
+    std::vector<char> intensity;
+    std::vector<char> alpha;
+  };
+  Pixels pixels;
```
Comment (pifon2a) [inline cartographer/io/submap_painter.h]: Is there some special reason, why this is not "std::vector<Pixel> pixels;", where
struct Pixel {
  char intensity;
  char alpha;
};
?

**Pull request merged**